src/Microsoft.AspNetCore.SignalR.Redis/RedisHubLifetimeManager.cs

```diff
@@ -199,34 +200,40 @@ public override Task SendGroupExceptAsync(string groupName, string methodName, o
                 throw new ArgumentNullException(nameof(groupName));
             }
 
-            var message = new RedisInvocationMessage(methodName, excludedIds, args);
-
-            return PublishAsync(_channelNamePrefix + ".group." + groupName, message);
+            var message = RedisProtocol.WriteInvocation(methodName, args, excludedIds, _hubProtocols);
+            return PublishAsync(_channels.Group(groupName), message);
         }
 
         public override Task SendUserAsync(string userId, string methodName, object[] args)
         {
-            var message = new RedisInvocationMessage(methodName, args);
-
-            return PublishAsync(_channelNamePrefix + ".user." + userId, message);
+            var message = RedisProtocol.WriteInvocation(methodName, args, _hubProtocols);
+            return PublishAsync(_channels.User(userId), message);
         }
 
-        private async Task PublishAsync(string channel, IRedisMessage message)
+        public override async Task AddGroupAsync(string connectionId, string groupName)
         {
-            byte[] payload;
-            using (var stream = new LimitArrayPoolWriteStream())
-            using (var writer = JsonUtils.CreateJsonTextWriter(new StreamWriter(stream)))
+            if (connectionId == null)
             {
-                _serializer.Serialize(writer, message);
-                writer.Flush();
-                payload = stream.ToArray();
+                throw new ArgumentNullException(nameof(connectionId));
+            }
+
+            if (groupName == null)
+            {
+                throw new ArgumentNullException(nameof(groupName));
             }
 
-            _logger.PublishToChannel(channel);
-            await _bus.PublishAsync(channel, payload);
+            var connection = _connections[connectionId];
+            if (connection != null)
+            {
+                // short circuit if connection is on this server
+                await AddGroupAsyncCore(connection, groupName);
```

Comment:
We want to avoid these optimizations IIRC, this leads to less testing of the other code paths when using redis on a single server.

Reply:
In this PR? This code was just moved around.